cpu/sam0_common/periph/usbdev.c

```diff
@@ -0,0 +1,739 @@
+
+static inline UsbDeviceEndpoint *_ep_reg_from_ep(usbdev_ep_t *ep)
+{
+    sam0_common_usb_t *usbdev = (sam0_common_usb_t *)ep->dev;
+    return &usbdev->config->device->DeviceEndpoint[ep->num];
+}
+
+static inline void _enable_irq(sam0_common_usb_t *dev)
+{
+    dev->config->device->INTENSET.reg = USB_DEVICE_INTENSET_EORST;
+}
+
+static inline void _disable_irq(sam0_common_usb_t *dev)
+{
+    dev->config->device->INTENCLR.reg = USB_DEVICE_INTENCLR_EORST;
+}
+
+static void _enable_ep_irq_out(sam0_common_usb_t *usbdev,
+                               UsbDeviceEndpoint *ep_reg)
+{
+    ep_reg->EPINTENSET.reg = USB_DEVICE_EPINTENSET_TRCPT0 |
+                             USB_DEVICE_EPINTENSET_STALL0 |
+                             USB_DEVICE_EPINTENSET_TRFAIL0;
+    if (ep_reg == &usbdev->config->device->DeviceEndpoint[0]) {
+        ep_reg->EPINTENSET.reg = USB_DEVICE_EPINTENSET_RXSTP;
+    }
+}
+
+static void _enable_ep_irq_in(UsbDeviceEndpoint *ep_reg)
+{
+    ep_reg->EPINTENSET.reg = USB_DEVICE_EPINTENSET_TRCPT1 |
+                             USB_DEVICE_EPINTENSET_STALL1 |
+                             USB_DEVICE_EPINTENSET_TRFAIL1;
+}
+
+/* Enable the endpoint specific interrupts */
+static void _enable_ep_irq(usbdev_ep_t *ep)
+{
+    UsbDeviceEndpoint *ep_reg = _ep_reg_from_ep(ep);
+
+    if (ep->dir == USB_EP_DIR_OUT) {
+        _enable_ep_irq_out((sam0_common_usb_t*)ep->dev, ep_reg);
+    }
+    else {
+        _enable_ep_irq_in(ep_reg);
+    }
+}
+
+static void _disable_ep_irq_out(sam0_common_usb_t *usbdev,
+                                UsbDeviceEndpoint *ep_reg)
+{
+    DEBUG("Disabling OUT irq\n");
+    ep_reg->EPINTENCLR.reg = USB_DEVICE_EPINTENCLR_TRCPT0 |
+                             USB_DEVICE_EPINTENCLR_STALL0 |
+                             USB_DEVICE_EPINTENCLR_TRFAIL0;
+    if (ep_reg == &usbdev->config->device->DeviceEndpoint[0]) {
+        ep_reg->EPINTENCLR.reg = USB_DEVICE_EPINTENCLR_RXSTP;
+    }
+}
+
+static void _disable_ep_irq_in(UsbDeviceEndpoint *ep_reg)
+{
+    DEBUG("Disabling IN irq\n");
+    ep_reg->EPINTENCLR.reg = USB_DEVICE_EPINTENCLR_TRCPT1 |
+                             USB_DEVICE_EPINTENCLR_STALL1 |
+                             USB_DEVICE_EPINTENCLR_TRFAIL1;
+}
+
+static void _bank_set_address(usbdev_ep_t *ep)
+{
+    UsbDeviceDescBank *bank = _bank_from_ep(ep);
+
+    bank->ADDR.reg = (uint32_t)ep->buf;
+}
+
+static int _bank_set_size(usbdev_ep_t *ep)
+{
+    UsbDeviceDescBank *bank = _bank_from_ep(ep);
+    unsigned val = 0x00;
+
+    switch (ep->len) {
+        case 8:
+            val = 0x0;
+            break;
+        case 16:
+            val = 0x1;
+            break;
+        case 32:
+            val = 0x2;
+            break;
+        case 64:
+            val = 0x3;
+            break;
+        case 128:
+            val = 0x4;
+            break;
+        case 256:
+            val = 0x5;
+            break;
+        case 512:
+            val = 0x6;
+            break;
+        case 1023:
```

Comment:
@bergzand I think I've read somewhere that some endpoint size can be 1023 or 1024 in specific case. Am I wrong ?

Reply:
1023 bytes is the max size for USB 2.0 full speed isochronous transfers, USB 2.0 high speed is allowed to have transfers of 1024 bytes with isochronous endpoint types.

Reply:
Alright then we're all good here :)